BeanCache findMap in combination with findOne does not work.

## Actual behavior

It seems that certain combinations with findMaps and findOnes are not properly handled by the bean cache.
(findList seems to work)
See the minimal example:

### Steps to reproduce

```java
  public void idsInFindMap() {
    List<OCachedBean> beans = createBeans(Arrays.asList("m0", "m1", "m2", "m3", "m4", "m5", "m6"));
    List<Long> ids = beans.stream().map(OCachedBean::getId).collect(Collectors.toList());
    beanCache.clear();
    beanCache.statistics(true);
    Query<OCachedBean> query = DB.find(OCachedBean.class).setUseCache(true);
    query.copy()
      .where().idIn(ids.subList(0, 1))
      .findMap(); // or findOne - it is important, that one entry is returned here

    query.copy()
      .where().idIn(ids.subList(0, 3))
      .findMap();

    query.copy()
      .where().idIn(ids.subList(2, 7))
      .findMap();
  }
```

```console
javax.persistence.PersistenceException: Error with property:3 dt:-5 data:6 java.lang.Long

	at io.ebeaninternal.server.persist.Binder.bindSimpleData(Binder.java:341)
	at io.ebeaninternal.server.persist.Binder.bindObject(Binder.java:234)
	at io.ebeaninternal.server.persist.Binder.lambda$bindObject$0(Binder.java:179)
	at io.ebeaninternal.server.persist.platform.MultiValueBind.bindMultiValues(MultiValueBind.java:50)
	at io.ebeaninternal.server.persist.Binder.bindObject(Binder.java:179)
	at io.ebeaninternal.server.expression.DefaultExpressionRequest.bind(DefaultExpressionRequest.java:62)
	at io.ebeaninternal.server.query.CQueryPredicates.bind(CQueryPredicates.java:139)
	at io.ebeaninternal.server.query.CQuery.prepareResultSet(CQuery.java:329)
	at io.ebeaninternal.server.query.CQuery.prepareBindExecuteQueryWithOption(CQuery.java:290)
	at io.ebeaninternal.server.query.CQuery.prepareBindExecuteQuery(CQuery.java:286)
	at io.ebeaninternal.server.query.CQueryEngine.findMany(CQueryEngine.java:344)
	at io.ebeaninternal.server.query.DefaultOrmQueryEngine.findMany(DefaultOrmQueryEngine.java:123)
	at io.ebeaninternal.server.core.OrmQueryRequest.findMap(OrmQueryRequest.java:432)
	at io.ebeaninternal.server.core.DefaultServer.findMap(DefaultServer.java:1130)
	at io.ebeaninternal.server.querydefn.DefaultOrmQuery.findMap(DefaultOrmQuery.java:1495)
	at io.ebeaninternal.server.expression.DefaultExpressionList.findMap(DefaultExpressionList.java:457)
	at org.tests.cache.TestBeanCache.idsInFindMap(TestBeanCache.java:76)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:725)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:214)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:210)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:135)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:66)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:107)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:53)
	at com.intellij.junit5.JUnit5IdeaTestRunner.startRunnerWithArgs(JUnit5IdeaTestRunner.java:57)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater$1.execute(IdeaTestRunner.java:38)
	at com.intellij.rt.execution.junit.TestsRepeater.repeat(TestsRepeater.java:11)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:35)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:235)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:54)
Caused by: org.h2.jdbc.JdbcSQLDataException: Unerlaubter Wert "3" für Parameter "parameterIndex"
Invalid value "3" for parameter "parameterIndex" [90008-214]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:646)
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:477)
	at org.h2.message.DbException.get(DbException.java:223)
	at org.h2.message.DbException.getInvalidValueException(DbException.java:298)
	at org.h2.jdbc.JdbcPreparedStatement.setParameter(JdbcPreparedStatement.java:1395)
	at org.h2.jdbc.JdbcPreparedStatement.setLong(JdbcPreparedStatement.java:678)
	at io.ebean.datasource.pool.ExtendedPreparedStatement.setLong(ExtendedPreparedStatement.java:223)
	at io.ebeaninternal.server.bind.DataBind.setLong(DataBind.java:129)
	at io.ebeaninternal.server.persist.Binder.bindSimpleData(Binder.java:271)
	... 85 more
```

--- a/ebean-core/src/main/java/io/ebeaninternal/server/core/DefaultServer.java
+++ b/ebean-core/src/main/java/io/ebeaninternal/server/core/DefaultServer.java
@@ -1107,12 +1107,13 @@ public final class DefaultServer implements SpiServer, SpiEbeanServer {
   @Override
   @SuppressWarnings({"unchecked", "rawtypes"})
   public <K, T> Map<K, T> findMap(Query<T> query, @Nullable Transaction transaction) {
-    SpiOrmQueryRequest request = createQueryRequest(Type.MAP, query, transaction);
+    SpiOrmQueryRequest request = buildQueryRequest(Type.MAP, query, transaction);
     request.resetBeanCacheAutoMode(false);
     if ((transaction == null || !transaction.isSkipCache()) && request.getFromBeanCache()) {
       // hit bean cache and got all results from cache
       return request.beanCacheHitsAsMap();
     }
+    request.prepareQuery();
     Object result = request.getFromQueryCache();
     if (result != null) {
       return (Map<K, T>) result;

--- a/ebean-test/src/test/java/org/tests/cache/TestBeanCache.java
+++ b/ebean-test/src/test/java/org/tests/cache/TestBeanCache.java
@@ -1,11 +1,12 @@
 package org.tests.cache;
 
-import io.ebean.xtest.BaseTestCase;
 import io.ebean.DB;
+import io.ebean.Query;
 import io.ebean.Transaction;
 import io.ebean.cache.ServerCache;
 import io.ebean.cache.ServerCacheStatistics;
 import io.ebean.test.LoggedSql;
+import io.ebean.xtest.BaseTestCase;
 import org.junit.jupiter.api.Test;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -55,9 +56,38 @@ public class TestBeanCache extends BaseTestCase {
   }
 
   @Test
+  public void idsInFindMap() {
+
+    List<OCachedBean> beans = createBeans(Arrays.asList("m0", "m1", "m2", "m3", "m4", "m5", "m6"));
+    List<Long> ids = beans.stream().map(OCachedBean::getId).collect(Collectors.toList());
+    beanCache.clear();
+    beanCache.statistics(true);
+    Query<OCachedBean> query = DB.find(OCachedBean.class).setUseCache(true);
+
+    // Test findIds
+    LoggedSql.start();
+    query.copy()
+      .where().idIn(ids.subList(0, 1))
+      .findMap(); // cache key is: 3/d[{/c1000}]/w[List[IdIn[?1],]]
+    assertThat(LoggedSql.stop().get(0)).contains("in (?)");
+
+    LoggedSql.start();
+    query.copy()
+      .where().idIn(ids.subList(0, 4))
+      .findMap(); // cache key is: 3/d[{/c1000}]/w[List[IdIn[?5],]]
+    assertThat(LoggedSql.stop().get(0)).contains("in (?,?,?,?,?)");
+
+    LoggedSql.start();
+    query.copy()
+      .where().idIn(ids.subList(2, 6))
+      .findMap(); // same cache key as above and same SQL above
+    assertThat(LoggedSql.stop().get(0)).contains("in (?,?,?,?,?)");
+  }
+
+  @Test
   public void idsIn_explicitCache_expect_cachePut() {
 
-    List<OCachedBean> beans = createBeans(Arrays.asList("k0","k1"));
+    List<OCachedBean> beans = createBeans(Arrays.asList("k0", "k1"));
     List<Long> ids = beans.stream().map(OCachedBean::getId).collect(Collectors.toList());
 
     beanCache.clear();
@@ -183,7 +213,7 @@ public class TestBeanCache extends BaseTestCase {
 
     Country country = DB.find(Country.class)
       .where()
-      .eq("name","NotValid")
+      .eq("name", "NotValid")
       .findOne();
 
     assertThat(country).isNull();

